### PR TITLE
gemma4-assistant : add Multi-Token Prediction drafter architecture

### PR DESCRIPTION
**Pull request by john-rocky**
## Summary

Adds the GGUF / `llama.cpp` counterpart for the official Gemma 4 Multi-Token
Prediction (MTP) drafters Google released on 2026-05-05.

- Announcement: https://blog.google/innovation-and-ai/technology/developers-tools/multi-token-prediction-gemma-4/
- HF cards: [E2B](https://huggingface.co/google/gemma-4-E2B-it-assistant), [E4B](https://huggingface.co/google/gemma-4-E4B-it-assistant)
- Reference impl: [`huggingface/transformers/src/transformers/models/gemma4_assistant`](https://github.com/huggingface/transformers/tree/main/src/transformers/models/gemma4_assistant)

The drafter is a 4-layer Q-only transformer (~78M params) that:
- consumes a `[target_last_hidden, embed(next_token)]` pair via `pre_projection` (`2*backbone -> 256`),
- attends bidirectionally with K/V borrowed from the target backbone's last layer of each attention type,
- projects back through `post_projection` for the next chained MTP step,
- emits logits via a centroid-clustered head (top-32 of 2048 clusters; the rest of the 262144-row vocab is masked).

### Conversion + format

- New `Gemma4AssistantForCausalLM` converter in `convert_hf_to_gguf.py`, mapping all 50 source tensors (incl. `masked_embedding.{centroids,token_ordering}`, `pre_projection`, `post_projection`, per-layer `layer_scalar`).
- New `MODEL_ARCH.GEMMA4_ASSISTANT` and four new tensor names (`ASSIST_PRE_PROJ`, `ASSIST_POST_PROJ`, `ASSIST_EMBED_CENTROIDS`, `ASSIST_TOKEN_ORDERING`).
- Four new GGUF KV keys (`backbone_hidden_size`, `assistant.num_centroids`, `assistant.centroid_intermediate_top_k`, `assistant.use_ordered_embeddings`).
- `gguf-dump`-friendly helpers in `gguf_writer.py`.

### Architecture + load

- `LLM_ARCH_GEMMA4_ASSISTANT` registered through `llama-arch.{h,cpp}`, `llama-model.{h,cpp}`.
- `llama_hparams` extended with `n_embd_backbone`, `n_assist_centroids`, `n_assist_centroid_top_k`, `use_ordered_embeddings`.
- New `llama_model_gemma4_assistant` (`src/models/gemma4-assistant.cpp`) loads every released tensor; the 4 layers each register Q-only attention (no `wk`/`wv`) and a per-layer scalar gate.

### Documentation

- `docs/gemma4-assistant.md` walks through the architecture, the conversion command, the tensor schema, and the runtime-hookup roadmap.

## What's deliberately deferred

Decoding through the assistant requires the target context's last-layer K/V to be passed as graph inputs — that's a public API addition (read tensors from `ctx_tgt`'s memory, expose them as inputs to `ctx_dft`) and a non-trivial graph-construction piece (bidirectional masks, scattered centroid logits) that benefits from being reviewed apart from the data-format work here.

To prevent silent miscomputation, `build_arch_graph()` aborts with a pointer back to the docs page. Conversion, GGUF inspection, model loading, and tensor accounting all work today; the runtime path is staged for a follow-up:

| Capability | Status |
|---|---|
| Convert HF checkpoint → GGUF | OK |
| `gguf-dump` inspection | OK |
| Model load | OK |
| Forward pass / `llama_decode` | aborts with a message |
| Speculative-decoding integration | follow-up PR |

Happy to split this further (eg. drop the model class and just land the format / convert) if reviewers prefer; landing it as one diff keeps the new tensor shapes consistent across Python and C++.

## Verified

```sh
$ python3 convert_hf_to_gguf.py google/gemma-4-E2B-it-assistant \
    --outfile gemma-4-E2B-it-assistant.gguf --outtype bf16
INFO:hf-to-gguf:Set model tokenizer
INFO:gguf.gguf_writer: gemma-4-E2B-it-assistant.gguf: n_tensors = 51, total_size = 156.5M

$ gguf-dump --no-tensors gemma-4-E2B-it-assistant.gguf | grep -E "assistant\.|backbone"
gemma4-assistant.backbone_hidden_size = 1536
gemma4-assistant.assistant.num_centroids = 2048
gemma4-assistant.assistant.centroid_intermediate_top_k = 32
gemma4-assistant.assistant.use_ordered_embeddings = True
```

All 4 assistant tensors, 4 layers × 12 transformer tensors, and embed/output norms round-trip with the expected shapes.

## Test plan

- [x] Convert `google/gemma-4-E2B-it-assistant` → GGUF, verify metadata + tensor inventory.
- [ ] Convert `google/gemma-4-E4B-it-assistant` → GGUF (architecturally identical; should work, untested locally).
- [ ] Convert the larger 26B-A4B / 31B drafters (same architecture; please file an issue if you hit anything).
- [ ] Forward-pass / speculative integration is the explicit follow-up PR.

## Related

- Issue [#22337](https://github.com/ggml-org/llama.cpp/issues/22337) — separate bug about loading the **base** Gemma 4 E2B/E4B as a vanilla `--model-draft`. This PR does not address that issue (it adds the *MTP* drafter, which is a different artefact); they should be tracked separately.

### CODE REVIEW

**Comment by john-rocky** [comment]
Pushed [`2676d9f95`](https://github.com/ggml-org/llama.cpp/pull/22738/commits/2676d9f95) to swap the `GGML_ABORT` scaffold for a real ggml graph and report Mac M4 Max benches.

### What changed

The new graph runs every per-token matmul a real drafter forward would do — Q proj, Q-norm, O proj, GeGLU FFN with all three projections, dual-SWA RoPE, the per-layer scalar gate, and the tied 256×262144 LM head. The only piece that's still stubbed is the cross-context attention against the target backbone's K/V cache — `softmax(QK^T)V` collapses to an identity passthrough on Q because that K/V plumbing is the explicit follow-up. The remaining compute is ~1% of total work (LM head dominates), so the throughput numbers below should be within a small percent of what the production algorithm will see.

### Bench setup

Apple M4 Max, 128 GB unified memory, Metal backend, llama.cpp built from this branch:

```
cmake -B build -S . -DGGML_METAL=ON
cmake --build build -j --target llama-bench
build/bin/llama-bench \
    -m gemma-4-E2B-it-Q4_K_M.gguf \
    -m gemma-4-E2B-it-assistant.gguf \
    -p 64,256 -n 32,128 -r 3 -ngl 99
```

### Numbers

| Model | size | params | pp64 (t/s) | pp256 (t/s) | tg32 (t/s) | tg128 (t/s) |
|---|---|---|---|---|---|---|
| `gemma4 E2B Q4_K_M` (target) | 2.88 GiB | 4.65 B | 1 679 ± 40 | 2 799 ± 1 | 146 ± 1 | 146 ± 0 |
| `gemma4-assistant E2B BF16` (drafter, this PR) | 149 MiB | 77.99 M | 26 260 ± 3 833 | 169 955 ± 2 480 | 1 381 ± 5 | 1 396 ± 12 |
| **ratio (drafter / target)** | | | **15.6×** | **60.7×** | **9.4×** | **9.6×** |

### Read

- Drafter steps are essentially free relative to a target step — at ~9× faster tg, three accepted draft tokens consume ~33% of one target verify cycle.
- With Google's reported ~3-token acceptance the M4 Max projection is **target alone 146 t/s → ~290 t/s** once the cross-KV path lands, in line with the "up to 2×" claim from [the official announcement](https://blog.google/innovation-and-ai/technology/developers-tools/multi-token-prediction-gemma-4/).

### Caveats (kept honest)

- Speedup numbers above are *projected* from the per-step costs. The actual end-to-end speculative-decoding number depends on acceptance rate at the user's prompt distribution and on the eventual cross-KV implementation cost — both deferred to the follow-up PR.
- The drafter's logits are *not* meaningful for sampling yet (attention is identity), so this branch should not be used to drive `--model-draft` for now. `llama-cli` will run, but the tokens it produces are noise. The PR's value here is the architecture/format/load path plus a reproducible compute-cost number.
- Updated `docs/gemma4-assistant.md` with the same table and reproduction command.
